src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -316,9 +336,92 @@ private string FormatEnumValueToString(string value, JavaScriptEncoder? encoder)
             return converted;
         }
 
+        private static string[] SplitFlagsEnum(string value)
+        {
+            // todo: optimize implementation here by leveraging https://github.com/dotnet/runtime/issues/934.
+            return value.Split(
+#if BUILDING_INBOX_LIBRARY
+                ValueSeparator
+#else
+                new string[] { ValueSeparator }, StringSplitOptions.None
+#endif
+                );
+        }
+
         internal override T ReadAsPropertyNameCore(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
+            bool success = TryParseEnumQuick(ref reader, options, out T value, out string? _);
```

Comment:
This should take `options.DictionaryKeyPolicy` into account (possibly overriding `_namingPolicy`). This might look like calling a helper method that looks like this:
```C#
bool TryParseEnum(ref reader, JsonNamingPolicy? namingPolicy, bool cacheResult, out T value);
```
and could be called here as follows:
```C#
bool cacheResult = true;
JsonNamingPolicy? namingPolicy = _namingPolicy;
if (options.DictionaryKeyPolicy != null)
{
    namingPolicy = options.DictionaryKeyPolicy;
    cacheResult = false; // no caching since multiple options can produce inconsistent results
}

bool success = TryParseEnum(ref reader, namingPolicy, cacheResult);
```
And the primary `Read` method:
```C#
bool success = TryParseEnum(ref reader, _namingPolicy, cacheResult: true);
```

Reply:
Per https://github.com/dotnet/runtime/pull/73348#discussion_r938178618, `DictionaryKeyPolicy` isn't used on deserialization:

https://github.com/dotnet/runtime/blob/dafc6dfd9dd0a38adc3ff10b0f178b3920151ce7/src/libraries/System.Text.Json/tests/Common/CollectionTests/CollectionTests.Dictionary.KeyPolicy.cs#L35.

Reply:
Got it, in that case we shouldn't change that right now, but it might make sense to create an issue and make things more consistent in the future. We should still remove the dictionary key policy cache per https://github.com/dotnet/runtime/pull/73348/files#r938627068

Reply:
It's by design - https://github.com/dotnet/runtime/issues/30142.

The dictionary key policy cache has been removed.

Reply:
Interesting, I wasn't aware of that issue. Going through James' points 1) and 2), I think that by caching all enum identifiers from the start, your change is careful to avoid calling `JsonNamingPolicy` for the deserialized values and has no need for an inverse `ConvertFrom` method. This is consistent with how we apply naming policies on object properties. Regarding point 3) your change is addressing that so I think we might as well bring back DictionaryKeyPolicy support.

Reply:
cc @JamesNK in the interest of transparency.

Reply:
I think I see why making `DictionaryKeyPolicy` deserialization work might be impractical: since the `options` instance is not tied to the converter, we cannot precompute the naming policy for all enum fields at construction time as we do with the converter-level naming policy. We would instead need to perform the same expensive calculation on each deserialization operation, which makes it infeasible to support in practice.